includes/abstracts/abstract-wc-order.php

```diff
@@ -983,22 +983,47 @@ public function remove_coupon( $code ) {
 	 * @since  3.2.0
 	 */
 	protected function recalculate_coupons() {
-		$discounts = new WC_Discounts( $this );
-		$coupons   = $this->get_items( 'coupon' );
-
-		foreach ( $coupons as $coupon ) {
-			$coupon_object = new WC_Coupon( $coupon->get_code() );
-			$discounts->apply_coupon( $coupon_object );
-		}
-
-		$this->set_coupon_discount_amounts( $discounts );
-
 		// Reset line item totals.
 		foreach ( $this->get_items() as $item ) {
 			$item->set_total( $item->get_subtotal() );
 			$item->set_total_tax( $item->get_subtotal_tax() );
 		}
 
+		$discounts = new WC_Discounts( $this );
+
+		foreach ( $this->get_items( 'coupon' ) as $coupon_item ) {
+			$coupon_code   = $coupon_item->get_code();
+			$coupon_id     = wc_get_coupon_id_by_code( $coupon_code );
+			$coupon_object = false;
+
+			if ( $coupon_id ) {
+				$coupon_object = new WC_Coupon( $coupon_id );
+			} elseif ( $original_coupon_data = $coupon_item->get_meta( 'coupon_data', true ) ) {
```

Comment:
Can we add comments on these if blocks? They make sense to me that the one is handling real and the other is handling virtual coupons but my future self might not see it immediately :)